### PR TITLE
Lint check

### PR DESCRIPTION
**Pull request by pducolin**
What does this PR do?
---------------------

Add go linter as a pre-commit hook and as a github action

Which scenarios this will impact?
-------------------

None

Motivation
----------

I accidentally added go errors in a previous PR, not merged yet. Protecting the repo from my mistakes

Additional Notes
----------------

### CODE REVIEW

File: .github/workflows/golangci-lint.yml [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+  # Optional: allow read access to pull request. Use with `only-new-issues` option.
+  pull-requests: read
+jobs:
+  golangci:
+    name: lint
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/setup-go@v3
+        with:
+          go-version: 1.17
```
Comment (vboulineau) [inline .github/workflows/golangci-lint.yml:20]: The Go version should match `go.mod` (1.19)

File: .github/workflows/golangci-lint.yml [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+      - uses: actions/checkout@v3
+      - name: golangci-lint
+        uses: golangci/golangci-lint-action@v3
+        with:
+          # Optional: version of golangci-lint to use in form of v1.2 or v1.2.3 or `latest` to use the latest version
+          version: v1.29
```
Comment (vboulineau) [inline .github/workflows/golangci-lint.yml:26]: Not sure we have an interest to have a fixed version (that we need to upgrade) in this repo.

File: .golangci.yaml [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+  # Allowed values: readonly|vendor|mod
+  # By default, it isn't set.
+  modules-download-mode: readonly
+  # Allow multiple parallel golangci-lint instances running.
+  # If false (default) - golangci-lint acquires file lock on start.
+  allow-parallel-runners: false
+  # Define the Go version limit.
+  # Mainly related to generics support since go1.18.
+  # Default: use Go version from the go.mod file, fallback on the env var `GOVERSION`, fallback on 1.18
+  go: '1.19'
```
Comment (vboulineau) [inline .golangci.yaml:55]: Not required, uses version in `go.mod`

File: .github/workflows/golangci-lint.yml (hunk 2 above)
Comment (pducolin) [reply .github/workflows/golangci-lint.yml:26]: Cons is, if a new linter adds new checks, we might find it blocking our PRs. Pros: we don't need to think about updates

File: .golangci.yaml [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+    - fmt.Printf
+    - fmt.Println
+    - (github.com/pulumi/pulumi/sdk/v3/go/pulumi.Log).Debug
+    - (github.com/pulumi/pulumi/sdk/v3/go/pulumi.Log).Info
+    - (github.com/pulumi/pulumi/sdk/v3/go/pulumi.Log).Warn
+    - (github.com/pulumi/pulumi/sdk/v3/go/pulumi.Log).Error
+    - (net/http).Write
+    - io.WriteString
+  staticcheck:
+    go: "1.17"
```
Comment (vboulineau) [inline .golangci.yaml:91]: Go version + probably too much copy paste from Agent repo.

Comment (pducolin) [reply .golangci.yaml:91]: For `staticcheck` you mean ? Should I rather remove all checks ?

Comment (pducolin) [reply .golangci.yaml]: Updated. Let me know if this works better !

**Review by vboulineau — APPROVED** [review]

**Pull request merged**